checker-rs/src/notebook_client.rs

```diff
@@ -0,0 +1,351 @@
+use crate::NotebookUser;
+
+use enochecker::{
+    tokio::{
+        io::{AsyncBufRead, AsyncBufReadExt, AsyncWriteExt, BufStream},
+        net::TcpStream,
+    },
+    CheckerError, CheckerRequest, CheckerResult,
+};
+use std::io;
+
+use tracing::{debug, error, info, instrument, warn};
+/// reads until the delimiter given in delim is found or EOF if found or another Error occurs
+///
+/// # Return
+/// Number of bytes read
+pub async fn read_until_slice<'a, T: AsyncBufRead + Unpin>(
+    stream: &'a mut T,
+    delim: &'a [u8],
+    buf: &'a mut Vec<u8>,
+) -> io::Result<usize> {
+    let mut bytes_read: usize = 0;
+    let mut inc_read;
+
+    let delim_end = *delim.last().expect("DELIMITER must not be empty");
```

Comment:
This should return `Err(InternalError)` instead, I assume - `expect` will lead to an empty result

Reply:
Returning a regular `io::Result<T>` makes it easier to work with.
I could return an `Err(io::Error(kind::Other, "Invalid Delimiter"))`, but calling read_until with no stop is sufficiently broken to warrant a panic imo.

Reply:
Nothing should ever panic inside `tokio` (as you told me), and building delimeters at runtime (and failing to do so) is a very valid usecase for this error. I would suggest returning `Err` here.
Especially if we move this to the lib at some point

Reply:
And I already checked with benni just closing the connection counts as a Checkererror which a panic does

Reply:
I agree it's gross misuse of the API - but still since this function already returns an error, why not do so here?

Reply:
>And I already checked with benni just closing the connection counts as a Checkererror

To be fair we should check that, it could also end up as down. An internal error would be more appropriate indeed, although this should become visible during tests with enochecker_test under realistic conditions.